### PR TITLE
[NU-1479] deployment status endpoint

### PR DESCRIPTION
**Pull request by arkadius**
## Checklist before merge
- [x] Related issue ID is placed at the beginning of PR title in \[brackets\] (can be GH issue or Nu Jira issue)
- [x] Code is cleaned from temporary changes and commented out lines
- [x] Parts of the code that are not easy to understand are documented in the code
- [x] Changes are covered by automated tests
- [ ] Showcase in dev-application.conf added to demonstrate the feature
- [ ] Documentation added or updated
- [ ] Added entry in _Changelog.md_ describing the change from the perspective of a public distribution user
- [ ] Added _MigrationGuide.md_ entry in the appropriate subcategory if introducing a breaking change
- [ ] Verify that PR will be squashed during merge

### CODE REVIEW

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentServiceNG.scala [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+      command: RunDeploymentCommandNG
+  ): EitherT[Future, RunDeploymentError, Future[Option[ExternalDeploymentId]]] = {
+    dbioRunner.runInTransactionE(
+      for {
+        scenarioMetadata <- scenariosRepository.getScenarioMetadata(command.scenarioName)
+        _ <- EitherT.fromEither(
+          Either.cond(command.user.can(scenarioMetadata.processCategory, Permission.Deploy), (), NoPermissionError)
+        )
+        _ <- deploymentRepository.saveDeployment(
+          DeploymentEntityData(command.id, scenarioMetadata.id, Timestamp.from(clock.instant()), command.user.id)
```
Comment (arkadius) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentServiceNG.scala:60]: I'm not sure about `command.user.id`. In other tables we use `command.user.username` and I'm pretty sure that it is bad. `username` causes that when some user changes the displayable name (e.g. surname part), in historical data there will be the outdated name. But on the other hand it is hard to predict when w fix that

**Comment by lukasz-bigorajski** [comment]
LGTM but I would wait for the @coutoPL opinion about errors in order to be on the same side in error handling.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    }
+  }
+
+  private def runDeployment(
+      command: RunDeploymentCommand
+  ): Future[Either[RunDeploymentError, Future[Option[ExternalDeploymentId]]]] = {
+    dbioRunner.runInTransactionE(
+      (for {
+        scenarioMetadata <- getScenarioMetadata(command)
+        _ <- EitherT.fromEither(
```
Comment (coutoPL) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala:56]: I can be simplified:
```scala
_ <- EitherT.cond(command.user.can(scenarioMetadata.processCategory, Permission.Deploy), (), NoPermissionError)
```

BTW, we could also create a `checkPermission` private method to make it more readable

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+      )
+    )
+  }
+
+  def getDeploymentStatus(
+      id: DeploymentId
+  )(implicit loggedUser: LoggedUser): Future[Either[GetDeploymentStatusError, StatusName]] =
+    dbioRunner.runInTransactionE(
+      (for {
+        deploymentWithScenarioMetadata <- EitherT[DB, GetDeploymentStatusError, DeploymentWithScenarioMetadata](
```
Comment (coutoPL) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala:115]: what about this?
```scala
deploymentWithScenarioMetadata <- EitherT(deploymentRepository.getDeploymentById(id))
```
Won't it work?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+          deploymentRepository.getDeploymentById(id)
+        )
+        DeploymentWithScenarioMetadata(_, scenarioMetadata) = deploymentWithScenarioMetadata
+        _ <- EitherT.cond(loggedUser.can(scenarioMetadata.processCategory, Permission.Read), (), NoPermissionError)
```
Comment (coutoPL) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala:119]: let's reuse the private method we've created (some comment above)

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        // TODO: We should check deployment state instead scenario state but before that we should pass correct deployment id
+        scenarioStatus <- EitherT.right[GetDeploymentStatusError](toEffectAll(DB.from {
```
Comment (coutoPL) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala:121]: can we move it to private, well-named method?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/description/DeploymentApiEndpoints.scala [hunk 6]
```diff
@@ -97,43 +139,43 @@ object DeploymentApiEndpoints {
...
-        Codec.string.map(RequestedDeploymentId(_))(_.value)
+    private def codec[Error](
+        toMessage: Error => String
+    ): Codec[String, Error, CodecFormat.TextPlain] =
+      Codec.string.map(
+        Mapping.from[String, Error](deserializationNotSupportedException)(toMessage)
+      )
 
-    }
+    private def deserializationNotSupportedException =
```
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/api/description/DeploymentApiEndpoints.scala:177]: Endpoints extends `BaseEndpointDefinitions` so I moved it to `BaseEndpointDefinitions`'s companion object. How do you think, is it a correct place?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/description/DeploymentApiEndpoints.scala [hunk 7]
```diff
@@ -97,43 +139,43 @@ object DeploymentApiEndpoints {
...
+    implicit def conflictingDeploymentIdErrorCodec: Codec[String, ConflictingDeploymentIdError, CodecFormat.TextPlain] =
+      codec[ConflictingDeploymentIdError](err => s"Deployment with id ${err.id} already exists")
 
-    object RequestedDeploymentId {
+    implicit def deploymentNotFoundErrorCodec: Codec[String, DeploymentNotFoundError, CodecFormat.TextPlain] =
+      codec[DeploymentNotFoundError](err => s"Deployment ${err.id} not found")
 
-      implicit val deploymentIdCodec: PlainCodec[RequestedDeploymentId] =
-        Codec.string.map(RequestedDeploymentId(_))(_.value)
+    private def codec[Error](
```
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/api/description/DeploymentApiEndpoints.scala:170]: You're right, I copy-pasted it. Now it is called `toTextPlainCodecSerializationOnly`. WDYT?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentRepository.scala [hunk 8]
```diff
@@ -0,0 +1,51 @@
+package pl.touk.nussknacker.ui.process.newdeployment
+
+import db.util.DBIOActionInstances._
+import pl.touk.nussknacker.ui.db.entity.ProcessEntityData
+import pl.touk.nussknacker.ui.db.{DbRef, NuTables}
+import pl.touk.nussknacker.ui.process.newdeployment.DeploymentEntityFactory.DeploymentEntityData
+import pl.touk.nussknacker.ui.process.newdeployment.DeploymentRepository.ConflictingDeploymentIdError
+import pl.touk.nussknacker.ui.process.newdeployment.DeploymentService.DeploymentNotFoundError
```
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentRepository.scala:8]: I changed the method singature to Option - one error mapping less ;-)

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/api/DeploymentApiHttpServiceBusinessSpec.scala [hunk 9]
```diff
@@ -23,29 +25,52 @@ class DeploymentApiHttpServiceBusinessSpec
...
+            .basicAuthAdmin()
+            .jsonBody(correctDeploymentRequest)
+            .put(s"$nuDesignerHttpAddress/api/deployments/$requestedDeploymentId")
+            .Then()
+            .statusCode(202)
+            .verifyApplicationState {
+              waitForDeploymentStatusMatches(requestedDeploymentId, SimpleStateStatus.Finished)
+            }
+            .verifyExternalState {
+              outputTransactionSummaryContainsExpectedResult()
```
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/test/scala/pl/touk/nussknacker/ui/api/DeploymentApiHttpServiceBusinessSpec.scala:53]: You're right. I though about it during writing it but I couldn't find the solution without too much code duplication or method names longer than method body ;-) I refactored it a bit, taka a look now. Still there is one not nice thing. There is no connection between expected results and input data in test. To make it more clear, we should move content of configuration (`tables-definition.sql`) and input directory (which is in resources) on the test level which become complicated for me. So I only left a comment. WDYTAT?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala (hunk 2 above)
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala]: Simplified. This private method would take: user, category, expectedPermission and it is one extra jump to understand what's going on. I think that it's an overhead ;-)

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala (hunk 3 above)
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala]: It is not relevant after changes around `DBIOActionRunner`

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala (hunk 5 above)
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala]: After changes in `DBIOActionRunner`, there is no `toEffectAll` and `DB.from`. Do you think that extracting a method for this oneliner still gives more benefits than losses?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/description/DeploymentApiEndpoints.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (coutoPL) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/api/description/DeploymentApiEndpoints.scala]: yeah, I thought about `BaseEndpointDefinitions` but for some reason I wrote `BaseHttpService`

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala (hunk 3 above)
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala]: For some reasons it didn't work but now it started to work :)

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala (hunk 5 above)
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala]: after reverting db.run related changes it become relevant again so I extracted this method

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentRepository.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (coutoPL) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentRepository.scala:8]: Do you? for some reason I don't see this change

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentRepository.scala [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+}
+
+object DeploymentRepository {
+
+  final case class ConflictingDeploymentIdError(id: DeploymentId)
+
+}
+
+case class DeploymentWithScenarioMetadata(deployment: DeploymentEntityData, scenarioMetadata: ProcessEntityData)
```
[the remark this answers is not in the document]
Comment (coutoPL) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentRepository.scala:51]: what about this one?

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/api/DeploymentApiHttpServiceBusinessSpec.scala [hunk 11]
```diff
@@ -23,29 +25,52 @@ class DeploymentApiHttpServiceBusinessSpec
...
-          .jsonBody(correctDeploymentRequest)
-          .put(s"$nuDesignerHttpAddress/api/scenarios/$scenarioName/deployments/$requestedDeploymentId")
-          .Then()
-          // TODO (next PRs): we should return 201 and we should check status of deployment before we verify output
-          .statusCode(200)
-          .verifyExternalState {
-            outputTransactionSummaryContainsResult()
-          }
+    "authenticated as user with deploy access" - {
+      "when invoked once should should" - {
```
[the remark this answers is not in the document]
Comment (coutoPL) [reply designer/server/src/test/scala/pl/touk/nussknacker/ui/api/DeploymentApiHttpServiceBusinessSpec.scala:42]: what about this one?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentRepository.scala (hunk 10 above)
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentRepository.scala:51]: Ehh, again GH hidden some comments, I'll check all them again

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentRepository.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentRepository.scala:8]: Woops I changed it only in `ScenarioMetadataRepository`, I'l change it also here

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/api/DeploymentApiHttpServiceBusinessSpec.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (coutoPL) [reply designer/server/src/test/scala/pl/touk/nussknacker/ui/api/DeploymentApiHttpServiceBusinessSpec.scala]: What about creating these files programmatically in `beforeAll` (and cleaning them in `afterAll`)? Using `betterfiles`, this should be sth really simple.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentRepository.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentRepository.scala]: fixed

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentRepository.scala (hunk 10 above)
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentRepository.scala]: done

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala (hunk 2 above)
Comment (coutoPL) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala]: IMO it's more readable:

```scala
  private def runDeployment(
      command: RunDeploymentCommand
  ): Future[Either[RunDeploymentError, Unit]] =
    dbioRunner.run(
      (for {
        scenarioMetadata <- getScenarioMetadata(command)
        _ <- checkPermission(user = command.user, category = scenarioMetadata.processCategory, permission = Deploy)
        _ <- saveDeployment(command, scenarioMetadata)
        runResult <- invokeLegacyRunDeploymentLogic(command, scenarioMetadata)
      } yield runResult).value
    )

  private def checkPermission(user: LoggedUser, category: String, permission: Permission) = {
    EitherT.cond[DB](user.can(category, Deploy), (), NoPermissionError)
  }
```

I like how the EitherT simplifies the for-comprehension, but it's sometimes too verbose, and encapsulating it in the private method in most cases helps to get rid of the noise which is not important for the logic.

But I leave it the decision to you.

File: designer/restmodel/src/main/scala/pl/touk/nussknacker/restmodel/BaseEndpointDefinitions.scala [hunk 12]
```diff
@@ -84,6 +84,16 @@ object BaseEndpointDefinitions {
...
 
+  def toTextPlainCodecSerializationOnly[T](
+      toMessage: T => String
+  ): Codec[String, T, CodecFormat.TextPlain] =
+    Codec.string.map(
+      Mapping.from[String, T](deserializationNotSupportedException)(toMessage)
+    )
+
+  private def deserializationNotSupportedException =
+    (_: Any) => throw new IllegalStateException("Deserializing errors is not supported.")
```
Comment (coutoPL) [inline designer/restmodel/src/main/scala/pl/touk/nussknacker/restmodel/BaseEndpointDefinitions.scala:95]: when you search for `throw new IllegalStateException("Des` you will find we have sth similar in several places. Could you please refactor the places? It looks like sth simple to do

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+        command.user.can(scenarioMetadata.processCategory, Permission.Deploy),
+        (),
+        NoPermissionError
+      )
+      _         <- saveDeployment(command, scenarioMetadata)
+      runResult <- invokeLegacyRunDeploymentLogic(command, scenarioMetadata)
+    } yield runResult).value
+
+  private def getScenarioMetadata(command: RunDeploymentCommand) =
+    EitherT[Future, RunDeploymentError, ProcessEntityData](
```
Comment (coutoPL) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala]: I mentioned great factory methods of the EitherT. Take a look:
```scala
    EitherT.fromOptionF(
      fopt = scenarioMetadataRepository.getScenarioMetadata(command.scenarioName),
      ifNone = ScenarioNotFoundError(command.scenarioName)
    )
```

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+  }
+
+  def getDeploymentStatus(
+      id: DeploymentId
+  )(implicit loggedUser: LoggedUser): Future[Either[GetDeploymentStatusError, StatusName]] =
+    (for {
+      deploymentWithScenarioMetadata <- EitherT(
+        dbioRunner.run(deploymentRepository.getDeploymentById(id))
+      )
+      DeploymentWithScenarioMetadata(_, scenarioMetadata) = deploymentWithScenarioMetadata
```
Comment (coutoPL) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala]: I'm not sure if this is needed. We can just access the field. Do we really need to unapply this?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala (hunk 2 above)
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala]: After wrapping the line, it looks pretty much the same ;-) But I changed it if it looks better for you.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala (hunk 4 above)
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala]: reused

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+                Left(NewCommentValidationError(msg))
+              }
+          }
+      )
+    )
+
+  def getDeploymentStatus(
+      id: DeploymentId
+  )(implicit loggedUser: LoggedUser): Future[Either[GetDeploymentStatusError, StatusName]] =
+    dbioRunner.run(
```
Comment (coutoPL) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala:110]: in this transaction, I see that only getting the deployment ID is a "real" DBIO action. Other things (permission check and getting scenario status) are lifted from Future. So maybe it's better to run the first thing and then "work" on the Futures level?

File: designer/restmodel/src/main/scala/pl/touk/nussknacker/restmodel/BaseEndpointDefinitions.scala (hunk 12 above)
Comment (arkadius) [reply designer/restmodel/src/main/scala/pl/touk/nussknacker/restmodel/BaseEndpointDefinitions.scala:95]: LOL, it has spread a lot, I only search by method name.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala (hunk 13 above)
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala]: Yeah I already used it in one place, now it is used here as well

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala (hunk 14 above)
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala]: I don't like long expressions like deploymentWithScenarioMetadata.scenarioMetadata but if you have some allergy to destructing assignment, I can change it :)

Comment (coutoPL) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala]: I general, no I don't. But I see a value in readability when the public method looks like this:
```
def businessMethod(...) {
   for {
     doSth
     checkSth
     writeSth
   } yield result
}
```

And when there is in the middle some form of distraction (I mean, sth technical), the readability is worse. I want to focus in this place on the algorithm (business logic).

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+            result
+              .map[Either[RunDeploymentError, Unit]](_ => Right(()))
+              .recover { case CommentValidationError(msg) =>
+                Left(NewCommentValidationError(msg))
+              }
+          }
+      )
+    )
+
+  def getDeploymentStatus(
```
Comment (coutoPL) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala:107]: let's move the public method above the private ones.

**Review by coutoPL — APPROVED** [review]
1. do we need some changelog entry?
2. [one missing comment](https://github.com/TouK/nussknacker/pull/5896#discussion_r1593593571)

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/api/DeploymentApiHttpServiceBusinessSpec.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/test/scala/pl/touk/nussknacker/ui/api/DeploymentApiHttpServiceBusinessSpec.scala]: Done. I didn't use betterfiles,  but with currently used `FileUtils` it looks pretty concise as well.

**Comment by arkadius** [comment]
> 1. do we need some changelog entry?
We need one changelog entry about experimental batch support. I'll add it in separate PR when I collect all related changes

> 2. [one missing comment](https://github.com/TouK/nussknacker/pull/5896#discussion_r1593593571)
Fixed

**Pull request merged**